sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -0,0 +1,123 @@
+    legacyTraverseAndSubstituteCTE(plan)
+  } else {
+    traverseAndSubstituteCTE(plan, false)
+  }
+
+  private def legacyTraverseAndSubstituteCTE(plan: LogicalPlan): LogicalPlan = {
+    plan.resolveOperatorsUp {
+      case With(child, relations) =>
+        // substitute CTE expressions right-to-left to resolve references to previous CTEs:
+        // with a as (select * from t), b as (select * from a) select * from b
+        relations.foldRight(child) {
+          case ((cteName, ctePlan), currentPlan) => substituteCTE(currentPlan, cteName, ctePlan)
+        }
+    }
+  }
+
+  /**
+   * Traverse the plan and expression nodes as a tree and replace matching references to CTE
+   * definitions.
+   * - If the rule encounters a WITH node then it substitutes the child of the node with CTE
+   *   definitions of the node right-to-left order as a definition can reference to a previous
+   *   one.
+   *   For example the following query is valid:
+   *   WITH
+   *     t AS (SELECT 1),
+   *     t2 AS (SELECT * FROM t)
+   *   SELECT * FROM t2
+   * - If a CTE definition contains an inner WITH node then substitution of inner should take
+   *   precedence because it can shadow an outer CTE definition.
+   *   For example the following query should return 2:
+   *   WITH
+   *     t AS (SELECT 1),
+   *     t2 AS (
+   *       WITH t AS (SELECT 2)
+   *       SELECT * FROM t
+   *     )
+   *   SELECT * FROM t2
+   * - If a CTE definition contains a subquery that contains an inner WITH node then substitution
+   *   of inner should take precedence because it can shadow an outer CTE definition.
+   *   For example the following query should return 2:
+   *   WITH t AS (SELECT 1 AS c)
+   *   SELECT max(c) FROM (
+   *     WITH t AS (SELECT 2 AS c)
+   *     SELECT * FROM t
+   *   )
+   * - If a CTE definition contains a subquery expression that contains an inner WITH node then
+   *   substitution of inner should take precedence because it can shadow an outer CTE
+   *   definition.
+   *   For example the following query should return 2:
+   *   WITH t AS (SELECT 1)
+   *   SELECT (
+   *     WITH t AS (SELECT 2)
+   *     SELECT * FROM t
+   *   )
+   * @param plan  the plan to be traversed
+   * @param inTraverse whether the current traverse is called from another traverse, only in this
+   *                   case name collision can occur
+   * @return then plan where CTE substitution is applied
+   */
+  private def traverseAndSubstituteCTE(plan: LogicalPlan, inTraverse: Boolean): LogicalPlan = {
+    plan.resolveOperatorsUp {
+      case With(child: LogicalPlan, relations) =>
+        val traversedChild = child transformExpressions {
+          case e: SubqueryExpression => e.withNewPlan(traverseAndSubstituteCTE(e.plan, true))
+        }
+
+        relations.foldRight(traversedChild) {
+          case ((cteName, ctePlan), currentPlan) =>
+            lazy val substitutedCTEPlan = traverseAndSubstituteCTE(ctePlan, true)
```

Comment:
we need this lazy? It seems all the tests below passes without this lazy...

Reply:
lazy is just a minor performance tuning here to make sure that `traverseAndSubstituteCTE(ctePlan, true)` runs only once even if `substituteCTE(...)` encounters `cteName` multiple times.

Reply:
We cannot do like this?
```
  private def substituteCTE(
      plan: LogicalPlan,
      cteName: String,
      ctePlan: LogicalPlan)
```
Anyway, I feel it is hard to read around here...

Reply:
Unfortunately, if I remove call by name parameter passing from `substituteCTE` then `lazy val substitutedCTEPlan` will be evaluated before calling `substituteCTE` with it.
I think both call by name passing of `ctePlan` and `lazy val substitutedCTEPlan` is required to make sure that:
- if `cteName` is never used then `traverseAndSubstituteCTE(ctePlan, true)` shouldn't run
- if `cteName` is used one or more times then `traverseAndSubstituteCTE(ctePlan, true)` should run only once

I added some comments to `traverseAndSubstituteCTE` to explain the logic there.

Reply:
Which test does fail for removing the call-by-name parameter?

Reply:
None of the tests fail actually, but both lazy and call-by-name are part of the optimization.